File: src/main/java/com/snackgame/server/applegame/domain/Coordinate.java
package com.snackgame.server.applegame.domain;

import com.snackgame.server.applegame.domain.exception.NegativeCoordinateException;

public class Coordinate {

    private final int y;
    private final int x;

    public Coordinate(int y, int x) {
        validateNonNegative(y);
        validateNonNegative(x);
        this.y = y;
        this.x = x;
    }

    private void validateNonNegative(int axisCoordinate) {
        if (axisCoordinate < 0) {
            throw new NegativeCoordinateException();
        }
    }

    public int getY() {
        return y;
    }

    public int getX() {
        return x;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;

        Coordinate that = (Coordinate)o;

        if (y != that.y)
            return false;
        return x == that.x;
    }

    @Override
    public int hashCode() {
        int result = y;
        result = 31 * result + x;
        return result;
    }
}

File: src/main/java/com/snackgame/server/game/snackgame/domain/Coordinate.kt
package com.snackgame.server.game.snackgame.domain

import com.snackgame.server.game.snackgame.exception.NegativeCoordinateException


class Coordinate(val y: Int, val x: Int) {

    init {
        validateNonNegative(y)
        validateNonNegative(x)
    }


    private fun validateNonNegative(axisCoordinate: Int) {
        if (axisCoordinate < 0) {
            throw NegativeCoordinateException()
        }
    }

    override fun equals(o: Any?): Boolean {

        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val that = o as Coordinate

        return if (y != that.y) false else x == that.x
    }

    override fun hashCode(): Int {
        var result = y
        result = 31 * result + x
        return result
    }

    override fun toString(): String {
        return "Coordinate{" +
                "y=" + y +
                ", x=" + x +
                '}'
    }
}
